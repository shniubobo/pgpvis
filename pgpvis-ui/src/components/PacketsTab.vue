<script setup lang="ts">
import { Tree, type TreeExpandedKeys } from "primevue";
import type { TreeNode } from "primevue/treenode";
import { computed, ref, watch } from "vue";

import {
  PublicKeyAlgorithmId,
  TypeId,
  type PacketSequence,
  type PublicKey,
  type PublicSubkey,
  type RsaEncryptSign,
  type UserId,
} from "pgpvis-core";

const props = defineProps<{ packets: PacketSequence }>();

const expanded_keys = ref<TreeExpandedKeys>({});

const packet_nodes = computed(() => {
  const packet_nodes: TreeNode[] = [];
  for (const [idx, packet] of props.packets.entries()) {
    if (packet.inner === undefined) {
      continue;
    }

    const format = packet.inner.header.inner.format;
    const type_id = packet.inner.header.inner.ctb.inner.type_id;
    const any_length = packet.inner.header.inner.length.inner;
    const length =
      any_length.encoding == "Full" || any_length.encoding == "Partial"
        ? `${any_length.length.toString()} (${any_length.encoding})`
        : "Indeterminate";
    const header_node: TreeNode = {
      key: `${idx}-0`,
      label: "Header",
      children: [
        {
          key: `${idx}-0-0`,
          label: "CTB",
          children: [
            {
              key: `${idx}-0-0-0`,
              label: `Format: ${format}`,
            },
            {
              key: `${idx}-0-0-1`,
              label: `Type ID: ${type_id}`,
            },
          ],
        },
        {
          key: `${idx}-0-1`,
          label: `Length: ${length}`,
        },
      ],
    };

    if (type_id === TypeId.PublicKey || type_id === TypeId.PublicSubkey) {
      const public_key = packet.inner.body.inner as PublicKey | PublicSubkey;
      const key_role = type_id === TypeId.PublicKey ? "Key" : "Subkey";
      const creation_time = public_key.creation_time.inner;
      // Creation time is stored as seconds as per RFC 9580, but here we need
      // milliseconds.
      const creation_time_iso = new Date(creation_time * 1000).toISOString();
      const algorithm = PublicKeyAlgorithmId[public_key.algorithm.inner];
      const key_material = public_key.key_material.inner;

      let key_material_nodes: TreeNode[] = [];
      if (public_key.algorithm.inner === PublicKeyAlgorithmId.RsaEncryptSign) {
        const n_length = (key_material as RsaEncryptSign).n.inner.length.inner;
        const e_length = (key_material as RsaEncryptSign).e.inner.length.inner;
        key_material_nodes = [
          {
            key: `${idx}-1-3-0`,
            label: "n",
            children: [
              {
                key: `${idx}-1-3-0-0`,
                label: `Length: ${n_length}`,
              },
              {
                key: `${idx}-1-3-0-1`,
                label: `Integers: [${n_length} octets]`,
              },
            ],
          },
          {
            key: `${idx}-1-3-1`,
            label: `e`,
            children: [
              {
                key: `${idx}-1-3-1-0`,
                label: `Length: ${e_length}`,
              },
              {
                key: `${idx}-1-3-1-1`,
                label: `Integers: [${e_length} octets]`,
              },
            ],
          },
        ];
      }

      const body_node: TreeNode = {
        key: `${idx}-1`,
        label: "Body",
        children: [
          {
            key: `${idx}-1-0`,
            label: `Version: ${public_key.version.inner}`,
          },
          {
            key: `${idx}-1-1`,
            label: `Creation Time: ${creation_time_iso} (${creation_time})`,
          },
          {
            key: `${idx}-1-2`,
            label: `Algorithm: ${algorithm}`,
          },
          {
            key: `${idx}-1-3`,
            label: "Key Material",
            children: key_material_nodes,
          },
        ],
      };

      packet_nodes.push({
        key: `${idx}`,
        label: `Public ${key_role} (${type_id}): ${public_key.key_id} (${algorithm})`,
        children: [header_node, body_node],
      });
    } else if (type_id === TypeId.UserId) {
      const user_id = packet.inner.body.inner as UserId;

      const body_node: TreeNode = {
        key: `${idx}-1`,
        label: "Body",
        children: [
          {
            key: `${idx}-1-0`,
            label: `User ID: ${user_id.user_id.inner}`,
          },
        ],
      };

      packet_nodes.push({
        key: `${idx}`,
        label: `User ID (${type_id}): ${user_id.user_id.inner}`,
        children: [header_node, body_node],
      });
    }
  }
  return packet_nodes;
});

watch(packet_nodes, expand_all_nodes);

function expand_all_nodes() {
  for (const node of packet_nodes.value) {
    expand_node(node);
  }
}

function expand_node(node: TreeNode) {
  if (node.children?.length) {
    expanded_keys.value[node.key] = true;

    for (const child of node.children) {
      expand_node(child);
    }
  }
}
</script>

<template>
  <Tree
    v-model:expanded-keys="expanded_keys"
    :value="packet_nodes"
    class="gutter-stable h-screen overflow-auto p-4"
  />
</template>

<style lang="postcss">
@reference "../style.css";

.p-tree-node-content {
  @apply py-0;
}

.p-tree-node-toggle-button {
  @apply hidden;
}

.p-tree-node-children {
  @apply pl-[2ch];
}
</style>
