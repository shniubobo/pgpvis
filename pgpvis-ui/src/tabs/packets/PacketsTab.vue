<script setup lang="ts">
// The vscode plugin can understand `@pgpvis-core`, but `vue-tsc` won't, for
// some files. At the time of writing, those files include this one, `InputTab`
// and `NodeTree`. The reason is unknown.
//
// TODO: Figure out the reason.
import { Node } from "../../../../pgpvis-core/Cargo.toml";

import NodeTree from "./NodeTree.vue";

const props = defineProps<{ nodes: Node[] }>();
</script>

<template>
  <div class="h-screen overflow-auto p-4 gutter-stable">
    <NodeTree
      v-for="(node, idx) in props.nodes"
      :key="idx"
      :node="node"
      :depth="0"
    />
  </div>
</template>
