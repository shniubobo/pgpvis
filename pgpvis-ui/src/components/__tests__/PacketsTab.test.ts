import { render } from "@testing-library/vue";
import { describe, expect, test } from "vitest";

import PacketsTab from "../PacketsTab.vue";

describe.concurrent.skip("Nodes should match snapshots", () => {
  describe("13: User ID", () => {
    test("@", () => {
      const { container } = render(PacketsTab, {
        props: {
          packets: [
            {
              offset: 0,
              length: 25,
              inner: {
                header: {
                  offset: 0,
                  length: 2,
                  inner: {
                    format: "OpenPGP",
                    ctb: {
                      offset: 0,
                      length: 1,
                      inner: {
                        type_id: 13,
                      },
                    },
                    length: {
                      offset: 1,
                      length: 1,
                      inner: {
                        encoding: "Full",
                        length: 23,
                      },
                    },
                  },
                },
                body: {
                  offset: 2,
                  length: 23,
                  inner: {
                    user_id: {
                      offset: 2,
                      length: 23,
                      inner: "John <john@example.com>",
                    },
                  },
                },
              },
            },
          ],
        },
      });
      expect(container).toMatchSnapshot();
    });
  });
});
