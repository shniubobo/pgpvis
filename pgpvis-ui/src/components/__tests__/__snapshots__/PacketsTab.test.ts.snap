// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nodes should match snapshots > 13: User ID > @ 1`] = `
<div>
  <div
    class="p-tree p-component gutter-stable h-screen overflow-auto p-4"
    data-pc-name="tree"
    data-pc-section="root"
    pc1=""
  >
    <!---->
    <!---->
    <div
      class="p-tree-root"
      data-pc-section="wrapper"
    >
      
      
      <ul
        class="p-tree-root-children"
        data-pc-section="rootchildren"
        role="tree"
      >
        
        <li
          aria-expanded="false"
          aria-label="User ID (13): John <john@example.com>"
          aria-level="1"
          aria-posinset="1"
          aria-setsize="2"
          class="p-tree-node"
          data-pc-section="node"
          role="treeitem"
          tabindex="0"
        >
          <div
            class="p-tree-node-content"
            data-p-selectable="false"
            data-p-selected="false"
            data-pc-section="nodecontent"
          >
            <button
              class="p-tree-node-toggle-button"
              data-pc-section="nodetogglebutton"
              tabindex="-1"
              type="button"
            >
              
              <svg
                aria-hidden="true"
                class="p-icon p-tree-node-toggle-icon"
                data-pc-section="nodetoggleicon"
                fill="none"
                height="14"
                viewBox="0 0 14 14"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.38708 13C4.28408 13.0005 4.18203 12.9804 4.08691 12.9409C3.99178 12.9014 3.9055 12.8433 3.83313 12.7701C3.68634 12.6231 3.60388 12.4238 3.60388 12.2161C3.60388 12.0084 3.68634 11.8091 3.83313 11.6622L8.50507 6.99022L3.83313 2.31827C3.69467 2.16968 3.61928 1.97313 3.62287 1.77005C3.62645 1.56698 3.70872 1.37322 3.85234 1.22959C3.99596 1.08597 4.18972 1.00371 4.3928 1.00012C4.59588 0.996539 4.79242 1.07192 4.94102 1.21039L10.1669 6.43628C10.3137 6.58325 10.3962 6.78249 10.3962 6.99022C10.3962 7.19795 10.3137 7.39718 10.1669 7.54416L4.94102 12.7701C4.86865 12.8433 4.78237 12.9014 4.68724 12.9409C4.59212 12.9804 4.49007 13.0005 4.38708 13Z"
                  fill="currentColor"
                />
              </svg>
              
            </button>
            <!---->
            <span
              class="p-tree-node-icon"
              data-pc-section="nodeicon"
            />
            <span
              class="p-tree-node-label"
              data-pc-section="nodelabel"
            >
              
              User ID (13): John &lt;john@example.com&gt;
              
            </span>
          </div>
          <!---->
        </li>
        
      </ul>
      
      
    </div>
  </div>
</div>
`;
